### ISSUE TITLE
Inconsistencies in working with Authenticode signed executables

### ISSUE THREAD

**Issue by nefarius** [Ido-Moshe-Github/CiDllDemo#5]
First of all, thank you very much for this publication!

I couldn't resist playing around with it but I found some inconsistencies/issues/side-effects when using `CiValidateFileObject` against a custom executable signed with a DigiCert Authenticode certificate:

### Issuer

```
CN = DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1
O = DigiCert, Inc.
C = US
```

### Verification

`signtool verify` appears to be happy:

```
signtool verify /pa test.exe
File: test.exe
Index  Algorithm  Timestamp
========================================
0      sha1       Authenticode

Successfully verified: test.exe
```

### Test results

Depending on the following OS versions and TESTSIGNING states I got the following results:

OS | TESTSIGNING | Result
---|---|---
Windows 11 22H2 | ON | works, can parse various certificate and publisher details
Windows 11 22H2 | OFF | fails with `ERROR:policy info is empty`
Windows 10 22H2 | ON | works, can parse various certificate and publisher details
Windows 10 22H2 | OFF | fails with `STATUS_INVALID_IMAGE_HASH`

Anybody seen something similar? Is this function *supposed* to work with any Authenticode signed file or only on those with the Microsoft "secret sauce" trusted certificate chains?

Thanks & cheers

**EDIT:** to clarify; I am *assuming* this function is supposed to work with non-Microsoft-magic files as well, if this assumption is wrong, please rectify it for me.

**Comment by nefarius** [comment]
@MiroKaku pardon the ping, thought this might interest you as well 😬

**Comment by ByteWhite1x1** [comment]
. . . I opened a new issue. As I originally intended.

**Comment by dretax** [comment]
@nefarius I've seen your other project where you use the related code in some ways.
Did you manage to resolve this issue? Globalsign is returning STATUS_INVALID_IMAGE_HASH without testsigning indeed for 19045

**Comment by nefarius** [comment]
> @nefarius I've seen your other project where you use the related code in some ways. Did you manage to resolve this issue? Globalsign is returning STATUS_INVALID_IMAGE_HASH without testsigning indeed for 19045

In my particular case the issue was the certificate used. I originally had used an ECC certificate from DigiCert and after a bit of back and forth with their support (who surprisingly quite quickly figured out the issue, props there) I got a new RSA based one which worked. So it still works for purposes outside of the Microsoft certificate ecosystem, however the compatibility matrix if you will is quite narrow.

**EDIT:** I just realized that my initial post already mentioned testing with `RSA4096 SHA384`, hm... Guess I need to dig deeper in my memory what else changed 🤔

**Comment by dretax** [comment]
> > @nefarius I've seen your other project where you use the related code in some ways. Did you manage to resolve this issue? Globalsign is returning STATUS_INVALID_IMAGE_HASH without testsigning indeed for 19045
> 
> In my particular case the issue was the certificate used. I originally had used an ECC certificate from DigiCert and after a bit of back and forth with their support (who surprisingly quite quickly figured out the issue, props there) I got a new RSA based one which worked. So it still works for purposes outside of the Microsoft certificate ecosystem, however the compatibility matrix if you will is quite narrow.
> 
> **EDIT:** I just realized that my initial post already mentioned testing with `RSA4096 SHA384`, hm... Guess I need to dig deeper in my memory what else changed 🤔

Buggers, let me know, I'd really appreciate some insight on this one

**Comment by nefarius** [comment]
> > > @nefarius I've seen your other project where you use the related code in some ways. Did you manage to resolve this issue? Globalsign is returning STATUS_INVALID_IMAGE_HASH without testsigning indeed for 19045
> > 
> > 
> > In my particular case the issue was the certificate used. I originally had used an ECC certificate from DigiCert and after a bit of back and forth with their support (who surprisingly quite quickly figured out the issue, props there) I got a new RSA based one which worked. So it still works for purposes outside of the Microsoft certificate ecosystem, however the compatibility matrix if you will is quite narrow.
> > **EDIT:** I just realized that my initial post already mentioned testing with `RSA4096 SHA384`, hm... Guess I need to dig deeper in my memory what else changed 🤔
> 
> Buggers, let me know, I'd really appreciate some insight on this one

It definitely had something to do with the way the executable was signed, the validation code was fine, I remember that much.

**Comment by nefarius** [comment]
I remember now; I ended up abandoning the use of `CiValidateFileObject` because I couldn't figure it out 😢 So for now it will remain in the realms of Windows mysteries I'm afraid.

**Comment by ByteWhite1x1** [comment]
> I remember now; I ended up abandoning the use of `CiValidateFileObject` because I couldn't figure it out 😢 So for now it will remain in the realms of Windows mysteries I'm afraid.

Is your Digicert an EV one? Can you provide your test.exe that's signed with your certificate so I can test. CiValidateFileObject works for me in the kernel (Windows 11 22H3).

**Comment by nefarius** [comment]
> > I remember now; I ended up abandoning the use of `CiValidateFileObject` because I couldn't figure it out 😢 So for now it will remain in the realms of Windows mysteries I'm afraid.
> 
> Is your Digicert an EV one? Can you provide your test.exe that's signed with your certificate so I can test. CiValidateFileObject works for me in the kernel (Windows 11 22H3).

Yes, EV from DigiCert. Can provide you one later. What about Windows 10 22h2?

**Comment by nefarius** [comment]
@ByteWhite1x1 alrighty, here you go [nefconc.zip](https://github.com/user-attachments/files/15593288/nefconc.zip) this is the way it was signed:

```cmd
signtool sign /tr http://timestamp.digicert.com /td sha256 /fd sha256 /n "Nefarius Software Solutions e.U." .\nefconc.exe
```

Same way I sign all my public releases pretty much. Sources of the binary I attached [can be found here](https://github.com/nefarius/nefcon).

**Comment by ByteWhite1x1** [comment]
[+] ComputeAuthenticodeHash \Device\HarddiskVolume3\temp3\nefconc.exe
[+] ValidateCertificate \Device\HarddiskVolume3\temp3\nefconc.exe
[+] CiCheckSignedFile returned 0x00000000

[+] CiValidateFileObject returned 0x00000000 \temp3\nefconc.exe

Your file validates using the both undocumented API's. The function is the same on Win10 22H2; there is no reason why it would not work.

**Comment by nefarius** [comment]
> [+] ComputeAuthenticodeHash \Device\HarddiskVolume3\temp3\nefconc.exe [+] ValidateCertificate \Device\HarddiskVolume3\temp3\nefconc.exe [+] CiCheckSignedFile returned 0x00000000
> 
> [+] CiValidateFileObject returned 0x00000000 \temp3\nefconc.exe
> 
> Your file validates using the both undocumented API's. The function is the same on Win10 22H2; there is no reason why it would not work.

OK but have you tested it on 10 22H2 as well? I explicitly mentioned it because it failed my test.

It's been a while so I might just re-run it myself, maybe something wasn't done correctly in testing or in recording the results.

**Comment by ByteWhite1x1** [comment]
Hi.

[+] Windows 10 Build: 19045 (22H2)
[+] ComputeAuthenticodeHash \Device\HarddiskVolume3\temp3\nefconc.exe
[+] ValidateCertificate \Device\HarddiskVolume3\temp3\nefconc.exe
[+] CiCheckSignedFile returned 0x00000000 (OK)

The output is from my anti-malware driver. Is your EV certificate something that can also be used to sign a kernel driver?

**Comment by nefarius** [comment]
Thanks for confirming, maybe my testing was faulty, it's been a while. Maybe the fault was even in resolving the executable path incorrectly, I need to go back to that code and make sure.

Yes, my EV cert could in theory cross sign kernel binaries, if we were still in the past where that was a thing.

**Comment by ByteWhite1x1** [comment]
I use the CiCheckSignedFile in my project https://hexderef.com/anti-malware-with-memory-forensics because it works also with DLL's. As soon as I find time. I'll also test Windows server 2022.

I'll also need an EV (or someone legitimate with one) to sign my kernel driver. But for now test mode does the job.

**Comment by nefarius** [comment]
Well, an EV cert no longer does you any good since cross signing certificates all have expired and the most used recent Windows 10 and 11 kernels do not load EV-signed binaries, they accept Microsoft attestation signed binaries you submitted to the Microsoft Partner Portal using your EV as a token of authenticity that the binaries you submitted actually came from the holder of the EV. The EV alone has no power anymore other than being an ID badge. This gets more complicated on older OS's ofc. but those are no longer relevant in the grand scheme of production releases.

**Comment by nefarius** [comment]
> I use the CiCheckSignedFile in my project https://hexderef.com/anti-malware-with-memory-forensics because it works also with DLL's. As soon as I find time. I'll also test Windows server 2022.

I found the code associated with the original issue, if you want to have a look, maybe you see something obvious: https://gist.github.com/nefarius/eaa6a0881036cc46626b2824f12f9a00

**Comment by dretax** [comment]
> I use the CiCheckSignedFile in my project https://hexderef.com/anti-malware-with-memory-forensics because it works also with DLL's. As soon as I find time. I'll also test Windows server 2022.
> 
> I'll also need an EV (or someone legitimate with one) to sign my kernel driver. But for now test mode does the job.

Wrong, that's the whole point. Testmode is giving you delusions, the function does not work properly without testmode enabled. @nefarius and I discussed this in the past. I stumbled upon making the same mistake before, basically in test mode it worked flawless, but fails without It any time.

**Comment by dretax** [comment]
@ByteWhite1x1 I've contacted you through your website, let's test the CiCheckSignedFile with my cert if you have the time.

**Comment by nefarius** [comment]
Keep us plebs in the loop top pls 😝

**Comment by ByteWhite1x1** [comment]
> @ByteWhite1x1 I've contacted you through your website, let's test the CiCheckSignedFile with my cert if you have the time.

Hi. Sorry for the delay. I've been really busy and have had breaks. I've just tested your executable discussed in this post https://github.com/Ido-Moshe-Github/CiDllDemo/issues/5#issuecomment-2150424997 on bare metal with the CiCheckSignedFile and it worked just fine with your executable.

Windows 10 Pro 22H2 Build: 19045.4291

\Device\HarddiskVolume3\temp3\nefconc.exe 	
00000089	0.03783230	[+] CiCheckSignedFileWrapper	
00000090	0.03832940	[+] CiCheckSignedFile returned 0x00000000 (the signature is valid)

"C:\Program Files (x86)\Windows Kits\10\App Certification Kit\signtool.exe" verify /pa /all C:\temp3\nefconc.exe

Here is a screenshot proof: https://i.imgur.com/qZAEnhM.png

I currently have no time for more tests. CiCheckSignedFile works just fine for me also on the latest Windows 11 versions >= 22H2. Test signing was disabled. I used my DSE bypass to get the anti-malware driver loaded. Sorry for the grey hat method but it was the only way to test it.

Steps as follows:

1) GetDigestAlgorithm
2) ComputeAuthenticodeHash
3) CiCheckSignedFile 

It required a great deal of effort to get all (3) steps working in a kernel driver. I can provide the source code for all (3) steps as a software work.

**Comment by arlenyan** [comment]
@ByteWhite1x1 would you mind sharing your source code? I'd like to attestation-sign a driver with it and test it on a host with testsigning off.

**Comment by arlenyan** [comment]
I saw in this [thread](https://community.osr.com/t/driver-siging-using-ecc-token/59145/4) a user BoPa reported:

>It is not limited to driver verification; it also cannot verify the signature of any ECC-signed PE file using CiValidateFileObject().
>
>After analyzing CI.dll, I found a function named MinCrypt_DisableEcdsa that sets a specific byte during execution. This function is called within CipInitialize, and the byte it sets is involved in signature verification. If this byte is set, the function returns STATUS_INVALID_IMAGE_HASH. I reset this byte using WinDbg and tested again, and everything worked fine.

**Comment by nefarius** [comment]
> It is not limited to driver verification; it also cannot verify the signature of any ECC-signed PE file using CiValidateFileObject().

I experienced this very issue myself like two years or so ago, I contacted DigiCert support and we switched from ECC to RSA4096 for my EV and this error went away, very nice to see why and that I didn't just imagine it. Also using `STATUS_INVALID_IMAGE_HASH` for multiple different error types is quite something 😒

**Comment by nefarius** [comment]
> [@ByteWhite1x1](https://github.com/ByteWhite1x1) would you mind sharing your source code? I'd like to attestation-sign a driver with it and test it on a host with testsigning off.

I second this, much time has passed and I just can't find the time to run through my entire VM test battery once more to make sure I didn't just misremember cases 😅

**Comment by ByteWhite1x1** [comment]
I am able to verify the digital signature of any PE file (.EXE, .DLL, .SYS) at the kernel level in real-time. So, for what purpose do you need these codes? Why use the ECC algorithm when RSA-4096 works well?

**Comment by nefarius** [comment]
The purpose is to cross-check and verify these claims and figure out where other people's implementation goes wrong?

**Comment by nefarius** [comment]
> Why use the ECC algorithm when RSA-4096 works well?

because that was dictated by the CA, not the one using the signing certificate? Not everyone has the option I'd guess.

**Comment by arlenyan** [comment]
> I am able to verify the digital signature of any PE file (.EXE, .DLL, .SYS) at the kernel level in real-time. So, for what purpose do you need these codes? Why use the ECC algorithm when RSA-4096 works well?

@ByteWhite1x1 I was hoping to take you up on your offer of providing your source code to potentially learn from your approach/implementation.

I have otherwise essentially arrived at the same conclusion as @nefarius:

`CiValidateFileObject` seems to work only if `testsigning` is `on`. 

`CiCheckSignedFile` as an alternative api works regardless of `testsigning`, but requires calculating the Authenticode hash. e.g. traverse the portable-executable format, read the dos header to find the PE file header to find the optional header to find the security data-directory that contains the hash/signing-certificate, then sha1 hash the portable-executable minus the checksum, certificate table, and other optional data-directories.

Btw @nefarius provided an excellent repo in [Domito](https://git.nefarius.at/nefarius/Domito) both in code and aggregating educational Authenticode resources. Thanks for that!

**Comment by ByteWhite1x1** [comment]
CiValidateFileObject works whether testsigning is on/off. AppId.sys exports it but  https://asec.ahnlab.com/en/62668/ (the reason I've implemented "Secure IOCTL" in my driver making those exploits impossible) MS stopped using it, I guess. In my tests, there is a memory leak in CiValidateFileObject, so it cannot be used for this reason either. And it only works with EXEs.

Yes I've fully working CiCheckSignedFile implementation for the kernel. (Basically SmartScreen that can also validate in real-time  loaded DLL's). It validates every PE file so far. What's your Telegram?

>CiCheckSignedFile as an alternative api works regardless of testsigning, but requires calculating the Authenticode hash. e.g. traverse >the portable-executable format, read the dos header to find the PE file header to find the optional header to find the security data->directory that contains the hash/signing-certificate, then sha1 hash the portable-executable minus the checksum, certificate table, >and other optional data-directories.

I spent at least a month of my time on this code, so these codes won't be easily given away by me. 4,999 USD in BTC. Thanks.
I am a mad busy and no time for any free test at the moment. I work on an anti-malware similar to AhnLab V3. 

As a side question. Does your IDA PRO 7.6 SP1 or newer "ida64.exe" has an embedded digital signature when you right-click on the executable > Properties?

**Comment by nefarius** [comment]
I don't have any crypto but I understand the desire for compensation, especially on something as annoyingly specialised as this. Maybe all participants here can join forces and figure something out?

**Comment by live3v1l** [comment]
Here's solution, it works on clean Windows 10 without TESTSIGNING ON.
status = CiValidateFileObject(FileObject,
                                    0x00000020, // magic flag
                                    0,
                                    &policyInfo,
                                    &timeStampPolicyInfo,
                                    &signingTime,
                                    thumbprint,
                                    &thumbprintSize,
                                    &thumbprintAlgorithm);

Tested on build 19041 and 19045.
I spent 5 hours of my time on it and I am happy it wasn't a month.
I don't want your bitcoins but I appreciate virtual hugs

**Comment by nefarius** [comment]
Brilliant find @live3v1l thank you for sharing! I'll give it a test as soon as time allows for it 🥳 

![Hug](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExdzdqYm53d3M1YzR2Z3FkMjE5NTZ5Y2RmajBhcXplY3N1ODgzNnowMiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/9d3LQ6TdV2Flo8ODTU/giphy.gif)

**Comment by dretax** [comment]
> Here's solution, it works on clean Windows 10 without TESTSIGNING ON. status = CiValidateFileObject(FileObject, 0x00000020, // magic flag 0, &policyInfo, &timeStampPolicyInfo, &signingTime, thumbprint, &thumbprintSize, &thumbprintAlgorithm);
> 
> Tested on build 19041 and 19045. I spent 5 hours of my time on it and I am happy it wasn't a month. I don't want your bitcoins but I appreciate virtual hugs

@nefarius and I had billions of time wasted on this. Appreciate it.

**Comment by dretax** [comment]
@nefarius indeed the magic value was the key. 0x00000020

It's a meme that some guy would throw bitcoin over this meme that we didn't notice through the reversal.
You may close the ticket.

`ValidateCert: Validating file \Program Files\WindowsApps\Microsoft.WindowsNotepad_11.2501.31.0_x64__8wekyb3d8bbwe\Notepad\Notepad.exe
ValidateCert: CiValidateFileObject returned 0x00000000
IsValidCert: Cert: size - 1918, algorithm - SHA256, publisher - O=Microsoft Corporation, L=Redmond, S=Washington, C=US, subject - Microsoft Code Signing PCA 2011
IsValidCert: Nothing proper found in publisher name.
IsValidCert: Cert: size - 1528, algorithm - SHA256, publisher - O=Microsoft Corporation, L=Redmond, S=Washington, C=US, subject - Microsoft Corporation
IsValidCert: Nothing proper found in publisher name.`
